Detect menu items invoked via keybinding and dispatch keybinding on renderer

alexandrudima opened the pull request:
Fixes #35225

This PR uses the observation from https://github.com/Microsoft/vscode/issues/35225#issuecomment-436271235 , that Electron gives an event when a menu item is invoked which contains the information about modifier key status. Through a heuristic (that the modifiers must match) we execute, only on mac, a separate code path where the keybinding is sent to the renderer and it is there interpreted using the keybinding service.

bpasero commented:
I resolved the merge conflict but let @sbatten handle this PR for inclusion as the owner of menus.

sbatten commented:
I think this will work well for the resolution of keybindings when not using the menu, but the menu itself will still show the keybinding it resolves against the global context and the enablement of that item will also be resolved against the global context.

alexandrudima commented on src/vs/platform/menubar/electron-main/menubar.ts:
```diff
@@ -621,17 +630,49 @@ export class Menubar {
 			}
 
-			this.runActionInRenderer(commandId);
+			if (isMacintosh && userSettingsLabel && Menubar._menuItemIsTriggeredViaKeybinding(event, userSettingsLabel)) {
```
:+1:

alexandrudima commented on src/vs/platform/keybinding/common/abstractKeybindingService.ts:
```diff
@@ -156,10 +156,20 @@ export abstract class AbstractKeybindingService extends Disposable implements IK
 
+	public dispatchByUserSettingsLabel(userSettingsLabel: string, target: IContextKeyServiceTarget): void {
+		const keybindings = this.resolveUserBinding(userSettingsLabel);
+		if (keybindings.length === 1) {
```
This can happen when multiple key combinations produce the same character. I've pushed a change to cover that case.

alexandrudima commented:
Thanks @sbatten for the review. I've addressed your comments.

---

> I think this will work well for the resolution of keybindings when not using the menu, but the menu itself will still show the keybinding it resolves against the global context and the enablement of that item will also be resolved against the global context.

Sure. I suggest we create another issue to track that. The original issue #35225 complains about the key bindings and not about the menu displaying the keybindings. I think they are different problems with different severity and I think they should be tackled independently.

sbatten commented:
@alexandrudima yes, this can be merged

jesaerys commented:
@alexdima I'm seeing a regression on this, but #35225 is locked and I can't comment. Should I open a new issue?

alexdima commented:
@jesaerys Yes, please open a new issue.
